### ISSUE TITLE
NeoHDF5 files created with Python 2 cannot be read with Python 3

### ISSUE THREAD

**Issue by apdavison** [NeuralEnsemble/python-neo#223]
This is because `_f_getAttr("_type")` is read as a bytes type, not string.

**Comment by titipata** [comment]
I have the same problem. Do you guys have any progress on this issue?

**Comment by legouee** [comment]
Hello,  

Thanks to the new version of neo 0.5, it is now possible with Python 3 to read NeoHDF5 files, created from Python 2, with neo 0.4 and neo 0.5.

To verified it, I have created an HDF5 file dset_Python2_Neo_04.h5 and a datasets /dset_Python2_Neo_04 in it, like this :

	import h5py
	import numpy as np
	#
	# Create a new file using defaut properties.
	#
	file = h5py.File('/dset_Python2_Neo_04.h5','w')
	#                  
	# Create a dataset under the Root group.
	#
	dataset = file.create_dataset("dset_Python2_Neo_04",(4, 6), h5py.h5t.STD_I32BE)
	print("Dataset dataspace is", dataset.shape)
	print("Dataset Numpy datatype is", dataset.dtype)
	print("Dataset name is", dataset.name)
	print("Dataset is a member of the group", dataset.parent)
	print("Dataset was created in the file with Python 2 and NEO 0.4", dataset.file)
	#
	# Initialize data object with 0.
	#
	data = np.zeros((4,6))
	#
	# Assign new values
	#
	for i in range(4):
	    for j in range(6):
	        data[i][j]= i*6+j+1	 
	#
	# Write data
	#
	print("Writing data...")
	dataset[...] = data
	#
	# Close the file before exiting
	#
	file.close()


Then , I readen with Python 3 the file obtained 

	import h5py
	import numpy as np
	#
	# Open an existing file using default properties.
	#
	file = h5py.File('/dset_Python2_Neo_04.h5','r+')
	#                 
	# Open "dset_Python2_Neo_04" dataset under the root group.
	#
	dataset = file['/dset_Python2_Neo_04']
	#
	# Read data back and print it.
	#
	print("Reading data back...with Python 2 NEO 0.4")
	data_read = dataset[...]
	print("Printing data...")
	print(data_read)
	#
	# Close the file before exiting
	#
	file.close()


The NeoHDF5 file created with Python 2, regardless of which version of neo is used, is correctly read by Python 3. I get the following result :

Reading data back...with Python 2 NEO 0.4
Printing data...
[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]]

**Comment by apdavison** [comment]
Hi @legouee 
Unfortunately this test, by using h5py directly, does not use the `neo.io.NeoHdf5IO` class to create or read the file.  To test this, you will need to create (with Python 2 and Neo 0.4) a neo `Block` containing some data (`AnalogSignals` and `SpikeTrains`) and then create the HDF5 file using something like:
```python
io = NeoHdf5IO("dset_Python2_Neo_04.h5")
io.write(block)
```
You should then read it into Neo 0.5 under Python 3:
```python
io = NeoHdf5IO("dset_Python2_Neo_04.h5")
data = io.read()
```

**Comment by legouee** [comment]
Hello, 

I just created, with Python 2.7 and Neo 0.4, a HDF5 file. To do this, I created a neo block containing three segments with random data from AnalogSignals and SpikeTrain. Here is the code I use:


import sys
import numpy as np
import tables as tb
import neo
from neo.io import NeoHdf5IO
from neo.core import Block,Segment,AnalogSignalArray, SpikeTrain, AnalogSignal
from quantities import kHz, nA

#
# create a Block with 3 Segment
#

blk = Block()

for ind in range(3):
    seg = Segment(name='segment %d' % ind, index=ind)
    blk.segments.append(seg)
    
#
# Populate the Block with AnalogSignalArray and SpikeTrain objects
#

for seg in blk.segments:
    # AnalogSignal
    a = AnalogSignalArray(np.random.randn(10000, 64)*nA, sampling_rate=10*kHz)
    seg.analogsignalarrays.append(a)
    # SpikeTrain
    train0 = SpikeTrain(times=[.01, 3.3, 9.3], units='sec', t_stop=10)
    seg.spiketrains.append(train0)


#
# write a NeoHdf5IO
#
io = NeoHdf5IO('dset_Python2_Neo_04.h5')
io.write(blk)


The file 'dset_Python2_Neo_04.h5' is created.
I now go under Python 3.6 and Neo 0.5 to read this file. I use the following code:


import neo
from neo.io import NeoHdf5IO

io = NeoHdf5IO('dset_Python2_Neo_04.h5')
data = io.read()


The h5 file is read correctly.

**Comment by apdavison** [comment]
Great. It looks as though this issue was resolved at some point in the Neo 0.5 development branch.
